tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -0,0 +1,265 @@
+        EXPECT_FALSE(ARE_FULL_HANDSHAKES(server_conn, client_conn));
+
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+    }
+
+    /* Test: A valid TLS1.3 ticket with TLS1.2 client falls back to a TLS1.2 connection */
+    {
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_NOT_NULL(server_conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate initial handshake to produce TLS1.3 session ticket */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(server_conn, client_conn));
+
+        uint8_t server_message[] = "server message";
+        uint8_t client_buffer[sizeof(server_message)];
+        s2n_blocked_status blocked = 0;
+
+        /* Send and receive some application data. This will ensure that the client actually receives the
+        * New Session Ticket post-handshake message. */
+        EXPECT_SUCCESS(s2n_send(server_conn, server_message, sizeof(server_message), &blocked));
+        EXPECT_SUCCESS(s2n_recv(client_conn, client_buffer, sizeof(client_buffer), &blocked));
+        EXPECT_BYTEARRAY_EQUAL(client_buffer, server_message, sizeof(server_message));
+
+        /* Prepare client and server for a second connection */
+        EXPECT_SUCCESS(s2n_connection_wipe(client_conn));
+        EXPECT_SUCCESS(s2n_connection_wipe(server_conn));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Client sets up a resumption connection with the received TLS1.3 session ticket data */
+        EXPECT_SUCCESS(s2n_connection_set_session(client_conn, cb_session_data, cb_session_data_len));
+
+        /* Set client config to TLS1.2 cipher preferences */
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls12_client_config));
+
+        /* Negotiate second connection */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        /* Falls back to TLS1.2 handshake */
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(server_conn, client_conn));
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS12);
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+    }
+
+    /* Test: A client with a valid TLS1.2 session ticket and TLS1.3 cipher preferences
+     * will fail when a TLS1.3 server tries to negotiate a TLS1.2 connection. */
```

Comment:
Include a short explanation of why-- if it confused you initially, it'll probably confuse the next person too.